src/ZkFold/Symbolic/Data/ByteString.hs

```diff
@@ -0,0 +1,130 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DerivingStrategies  #-}
+{-# LANGUAGE TypeApplications    #-}
+
+module ZkFold.Symbolic.Data.ByteString (
+    ByteString(..),
+) where
+
+import           Control.Monad                                             (forM, mapM, zipWithM)
+import           Data.Bits
+import           Data.List                                                 (map)
+import           Data.Proxy                                                (Proxy (..))
+import           GHC.TypeNats                                              (KnownNat, Natural, natVal)
+import           Prelude                                                   (Integer, error, ($), (.))
+import qualified Prelude                                                   as Haskell
+import           Test.QuickCheck                                           (Arbitrary (..))
+
+import           ZkFold.Base.Algebra.Basic.Class
+import           ZkFold.Base.Algebra.Basic.Field                           (Zp)
+import           ZkFold.Prelude                                            (replicate)
+import           ZkFold.Symbolic.Compiler
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Combinators    (expansion, horner)
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.MonadBlueprint (ClosedPoly)
+import           ZkFold.Symbolic.Data.Combinators
+import           ZkFold.Symbolic.Data.UInt
+
+newtype ByteString (n :: Natural) a = ByteString (UInt n a)
```

Comment:
`UInt n a` needs some additional space to perform integer arithmetic, `ByteString`s could be stored more efficiently

Reply:
I guess it is in the task description. But I agree that if we can pack more tightly, then we should.